java/arcs/core/allocator/Arc.kt

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.allocator
+
+import arcs.core.common.ArcId
+import arcs.core.host.ArcState
+import arcs.core.host.ArcState.*
+import arcs.core.util.Log
+import kotlinx.coroutines.CompletableDeferred
+
+/**
+ * Represents an instantiated Arc running on one or more [ArcHost]s. An [Arc] can be stopped
+ * via [Arc.stop], and best-effort state changes can be listened for via [onArcStateChange].
+ *
+ * NOTE: that currently [onArcStateChange] listeners are not persistent across service restarts,
+ * therefore if an [ArcHost] is running in a [Service], or on another device, and they restart,
+ * you will not receive updates from those hosts.
+ *
+ * TODO: add some mechanism to detect host crashes and re-register state change listeners.
+ */
+data class Arc internal constructor(
+    val arcId: ArcId,
+    private val allocator: Allocator,
+    private var arcStateInternal: ArcState = NeverStarted
```

Comment:
Do we need Allocator here? Seems we are using Allocator to perform certain actions on this Arc, which makes it no longer a data class to me. 

If you plan to make this the ultimate one class for clients to handle Arc related stuff, how about add a start() method too? So the code will be like:

Arc arc = Arc(...)
arc.start()
...
arc.stop()

Compared to what we have now:
Allocator allocator = Allocator()
Arc arc = allocator.startArc(...)
...
arc.stop()

Reply:
It's needed to get access to ArcHosts so it can add change listeners to them. However, I could change this pass the ArcHost references to Arc directly and remove Allocator. 

I don't think Arc should be the way you start Arcs (e.g.  not the control API), it's more or less the way to get info about already running Arcs. stop() doesn't need to be there, it's a convenience. 

But it does mean that you need to keep a reference around to the Allocator if you want to stop an arc. This isn't too bad if you have dependency injection tho, and probably better for memory leak issues if Allocator is not being held by Arc objects.

Reply:
The convenience could also bring confusion as I see the control API being spread between two things: Allocator and Arc, which confuses me about whether Arc is just a pure data class which records states or it can have internal logics. And if it has internal logics, I'm starting to worry if I can safely pass it around.

Reply:
I agree with Yuan here, it seems like it's exposing too much API surface to holders of an Arc object to have `Allocator` as one of its fields.